src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,168 @@
+
+    public ArrayList(int customCapacity) {
+        if (customCapacity > 0) {
+            this.elementData = new Object[customCapacity];
+        } else if (customCapacity == 0) {
+            this.elementData = EMPTY_ELEMENTDATA;
+        } else {
+            throw new RuntimeException("Impossible capacity: "
+                    + customCapacity);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = grow(size + 1);
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        if (size == elementData.length) {
+            elementData = grow(size + 1);
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] a = list.toArray();
+        int addLength = a.length;
+        if (addLength == 0) {
+            return;
+        }
+        final int s = size;
+        if (addLength > (this.elementData.length - s)) {
+            this.elementData = grow(s + addLength);
+        }
+        System.arraycopy(a, 0, this.elementData, s, addLength);
+        size = s + addLength;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T get(int index) {
-        return null;
+        rangeCheckForGet(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheckForGet(index);
+        elementData[index] = value;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheckForGet(index);
+        T oldValue = (T) elementData[index];
+        fastRemove(index);
+        return oldValue;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T remove(T element) {
-        return null;
+        final int index = findElement(element);
+        T oldValue = (T) elementData[index];
+        fastRemove(index);
+        return oldValue;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public String toString() {
+        return Arrays.toString(elementData);
+    }
+
+    @Override
+    public Object[] toArray() {
+        return Arrays.copyOf(elementData, size);
+    }
+
+    private Object[] grow(int minCapacity) {
+        return elementData = Arrays.copyOf(elementData, newCapacity(minCapacity));
+    }
+
+    private int newCapacity(int minCapacity) {
+        int oldCapacity = elementData.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        if (newCapacity - minCapacity <= 0) {
```

Comment:
Magic numbers should be constants. The growth factor (oldCapacity >> 1) should be defined as a constant with an informative name.